Try to avoid if/elsif chain for fun.

gwadej opened the pull request:
Just trying to something a little different than the obvious.

Commit:
Try to avoid if/elsif chain for fun.

jaybobo commented on letsdrill.rb:
```diff
@@ -1,12 +1,12 @@
- 
-  #Put your code here!
-
+  grade_levels = [ [ 90, "A" ], [80, "B"], [70, "C"], [60, "D"] ]
```
nice. using a hash could be another option here that may prove more readable.

```ruby
grade_levels = { 90 => "A", 80 => "B", 70 => "C", 60 => "D", 0 => "F" }
grade_levels.each { |score,letter| return letter if integer >= score }
```

gwadej replied:
I thought about that, but the algorithm only works if I'm working through the grades in decreasing order. I didn't think I could assume that while walking the hash.

jaybobo replied:
good point in this case I believe you can provided that you're not mutating it. 

there's also this too.. http://api.rubyonrails.org/classes/ActiveSupport/OrderedHash.html

jaybobo commented on letsdrill.rb:
```diff
@@ -1,12 +1,12 @@
 def shortest_string(array)
 
-  #Put your code here!
+  return nil if array.length == 0
```
i think you can do away with this check.

`array.sort_by(&:length).first` should return nil

gwadej commented:
I tried the hash form @jaybobo suggested, and I get "LocalJumpError: unexpected return". Do I have the syntax wrong, or am I just misunderstanding something?

Commit:
Change the shortest_string code to be more idiomatic

gwadej commented:
Took into account some of @jaybobo 's suggestions.

jaybobo commented:
@gwadej if you want to send me a code snippet I can take a look at that localjumperror for you.

gwadej commented:
@jaybobo I used the two lines you sent me in irb to test.

> grade_levels = { 90 => "A", 80 => "B", 70 => "C", 60 => "D", 0 => "F" }
> grade_levels.each { |score,letter| return letter if 77 >= score }

And my results were:

> LocalJumpError: unexpected return
> 	from (irb):2:in `block in irb_binding'
> 	from (irb):2:in `each'
> 	from (irb):2
> 	from /Users/wjohnson/.rbenv/versions/2.0.0-p647/bin/irb:12:in `<main>'

jaybobo commented:
@gwadej Try wrapping that in a method. Blocks behave differently than proc or lambdas.

gwadej commented:
Okay. That worked. It seems to give the right letter each time.
The Hash#each documentation does not appear to guarantee the order of the iteration. The old copy of the PickAxe book that I've got says that the iterators return the items in an arbitrary order. That matches my understanding of the underlying hash table.

It appears that the PickAxe 1.9/2.0 says the items will *generally be the order that the entries were initially inserted into the hash*.

That _generally_ kind of worries me. How do I know when the iteration order matches insertion order?

jaybobo commented:
@gwadej yeah great point, i'd love to get a better understanding of _generally_ here as well. at what point does it _generally not_ preserve order of insertion ha? 

ActiveSupport::OrderedHash is the safest bet here.

Commit:
Minor tweaks to the search function.

Document reason for using Array over Hash.
Really not comfortable with relying on behavior that does not appear to
be guaranteed and radically changed in a recent version.

gwadej commented:
It appears that ActiveSupport::OrderedHash just uses the Hash behavior internally unless the version is less than 1.9.

This whole topic seems to be a potential source of version-specific bugs.
